### PR TITLE
108 feature request add test data loop functionality along with trainval to 2d classifier

### PR DESCRIPTION
**Pull request by mikewoodward94**
https://github.com/GSTT-CSC/project-template/issues/108

Solve the above issue.

1) Adds validation and test splits to the config
2) Update datamodule to make these splits and add dataset/loader for test data
3) Add metrics to be completed during test_step
4) Evaluate best model also runs for test data.

Tune.py shouldn't use the test data at all.

### CODE REVIEW

File: project/classifier_2d/src/datamodule.py [hunk 1]
```diff
@@ -85,17 +86,28 @@ def setup(self, stage: Optional[str] = None):
 
         self.labels = [int(sample['action_data']) for subject in data for sample in subject['data'] if
                        sample['data_label'] == 'label']
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/datamodule.py:88]: Could you use get_labels() for this snippet here too? I see you use it now for train_val_labels so might as well use it here too.

File: project/classifier_2d/src/datamodule.py [hunk 2]
```diff
@@ -85,17 +86,28 @@ def setup(self, stage: Optional[str] = None):
...
-                                                                 stratify=self.labels,
-                                                                 random_state=self.random_seed,
-                                                                 )
+        
+        train_val_data, self.test_data = train_test_split(
+            data,
+            test_size=self.test_fraction,
+            stratify=self.labels,
+            random_state=self.random_seed,
+            )
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/datamodule.py:95]: This will (annoyingly) fail if test_size = 0 - you'll probably need to adjust the logic to return something like None for self.test-dataset if test_size = 0, and then skip the testing part in Network.

File: project/classifier_2d/src/datamodule.py [hunk 3]
```diff
@@ -85,17 +86,28 @@ def setup(self, stage: Optional[str] = None):
...
+            test_size=val_size,
+            stratify=train_val_labels,
+            random_state=self.random_seed,
+            )
 
         self.data_manifest = {
             'N_total': len(data),
             'train': self.dataset_stats(self.train_data),
             'validation': self.dataset_stats(self.validation_data),
+            'test': self.dataset_stats(self.test_data),
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/datamodule.py:110]: probably also have something like `self.dataset_stats(self.test_data) if self.test_data else None` or something

File: project/classifier_2d/src/datamodule.py
Comment (mikaelsimard5) [inline project/classifier_2d/src/datamodule.py:116]: This bit also needs to accommodate the case where there is no self.test_data!

File: project/classifier_2d/src/network.py [hunk 4]
```diff
@@ -230,6 +241,47 @@ def validation_step(self, batch, batch_idx):
 
         return {"loss": loss, 'y_onehot': y_onehot, 'y_pred_act': y_pred_act}
 
+    def test_step(self, batch, batch_idx):
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/network.py:244]: test_step is basically the same as val_step - suggest we could define a shared helper function in Network such as _eval_step and then call it both in val_step and test_step so that we only have one place where code needs to be modified? In _eval_step you can have a split="val" or "test" argument for logging.

File: project/classifier_2d/src/network.py [hunk 5]
```diff
@@ -103,12 +110,16 @@ def __init__(self,
                                 torch.tensor(train_class_weights, dtype=torch.float32))
             self.register_buffer('validation_class_weights_tensor',
                                 torch.tensor(validation_class_weights, dtype=torch.float32))
+            self.register_buffer('test_class_weights_tensor',
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/network.py:113]: also needs to mitigate test_class_weights = None or so

File: project/classifier_2d/src/network.py [hunk 6]
```diff
@@ -103,12 +110,16 @@ def __init__(self,
...
+                        torch.tensor(test_class_weights, dtype=torch.float32))
         else:
             self.train_class_weights_tensor = None
             self.validation_class_weights_tensor = None
+            self.test_class_weights_tensor = None
 
         self.train_loss_function = get_loss_function(loss_fcn, weight=self.train_class_weights_tensor, label_smoothing=self.label_smoothing)
         self.validation_loss_function = get_loss_function(loss_fcn, weight=self.validation_class_weights_tensor)
+        self.test_loss_function = get_loss_function(loss_fcn, weight=self.test_class_weights_tensor)
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/network.py:122]: Similarly, manage if test_size = 0

File: project/classifier_2d/mlops/tune.py
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/tune.py:1]: I don't really have a simple solution at the moment but imagine the following - 

if we use test_fraction = 0 in tune, we've used all of the data to find a model. If test_fraction > 0 in train, then we'll end up having data we already used for tuning in the test set which creates a biased result. 

The solution would naturally be - if you want to have a test set, then you should also remove this data from tune.py. This should be achievable by using the exact same dataset, test_fraction and seed. However this is not checked/enforced at all in train.py, and it's quite hard to make sure we're syncing this information across independent scripts.

In general I don't think it's much of a problem as we've been doing testing independently with a different dataset on XNAT. However if we include an option for testing in train.py then we need to make the user aware of this, otherwise we may end up with biased models!

File: project/classifier_2d/mlops/train.py [hunk 7]
```diff
@@ -129,6 +132,7 @@ def train(config):
         )
 
         trainer.fit(net, dm)
+        trainer.test(net, dm, ckpt_path=checkpoint_callback.best_model_path)
```
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/train.py:135]: also make sure this is managed with test size being 0

File: project/classifier_2d/mlops/train.py [hunk 8]
```diff
@@ -168,10 +172,12 @@ def train(config):
         logger.info('Finding best threshold')
         best_model_path = checkpoint_callback.best_model_path
         best_model = net.load_from_checkpoint(best_model_path)
-        net.evaluate_best_model(best_model, threshold_tune=True)
+        net.evaluate_best_model(best_model, split='val')
+        net.evaluate_best_model(best_model, split='test')
```
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/train.py:176]: also manage with test size = 0

File: project/classifier_2d/mlops/train.py [hunk 9]
```diff
@@ -168,10 +172,12 @@ def train(config):
...
 
         mlflow.log_artifact('val_transform_failures.csv')
         mlflow.log_artifact('train_transform_failures.csv')
+        mlflow.log_artifact('test_transform_failures.csv')
```
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/train.py:180]: test size = 0 etc :D

**Review by mikaelsimard5 — CHANGES_REQUESTED** [review]
Mostly minor to catch exceptions although I think I am raising an important issue (see the comment on tune.py) re the use of a test set in train.py
